nexus/db-queries/src/db/queries/vpc.rs

```diff
@@ -278,3 +270,79 @@ impl NextVni {
 }
 
 delegate_query_fragment_impl!(NextVni);
+
+// Helper type to compute the shift for a `NextItem` query to find VNIs.
+#[derive(Clone, Copy, Debug, PartialEq)]
+struct VniShifts {
+    // The minimum `ShiftGenerator` shift.
+    min_shift: i64,
+    // The maximum `ShiftGenerator` shift.
+    max_shift: i64,
+}
+
+/// Restrict the search for a VNI to a small range.
+///
+/// VNIs are pretty sparsely allocated (the number of VPCs), and the range is
+/// quite large (24 bits). To avoid memory issues, we'll restrict a search
+/// for an available VNI to a small range starting from the random starting
+/// VNI.
+//
+// NOTE: This is very small for tests, to ensure we can accurately test the
+// failure mode where there are no available VNIs.
+#[cfg(not(test))]
+pub const MAX_VNI_SEARCH_RANGE_SIZE: u32 = 2048;
+#[cfg(test)]
+pub const MAX_VNI_SEARCH_RANGE_SIZE: u32 = 10;
+
+impl VniShifts {
+    fn new(vni: Vni) -> Self {
+        let base_u32 = u32::from(vni.0);
+        let range_end = base_u32 + MAX_VNI_SEARCH_RANGE_SIZE;
+
```

Comment:
Should we add a https://docs.rs/static_assertions/latest/static_assertions/macro.const_assert.html that `MAX_VNI_SEARCH_RANGE_SIZE <= (MAX_VNI - MIN_GUEST_VNI)` ? Otherwise it seems like our wrap-around "min_shift" might collide with the part of the range we previously checked.

Reply:
Good idea, done in ca6d7ccdf